dev/TreeView/TreeViewNode.cpp

```diff
@@ -193,27 +193,43 @@ void TreeViewNode::OnItemsSourceChanged(const winrt::IInspectable& sender, const
 
 void TreeViewNode::OnItemsAdded(int index, int count)
 {
-    for (int i = index + count - 1; i >= index; i--)
+    // TreeViewNode and ItemsSource will update each other when data changes.
+    // For ItemsSource -> TreeViewNode changes, m_itemsDataSource.Count() > Children().Size()
+    // We'll add the new node to children collection.
+    // For TreeViewNode -> ItemsSource changes, m_itemsDataSource.Count() == Children().Size()
+    // the node is already in children collection, we don't want to update TreeViewNode again here.
+    if (m_itemsDataSource.Count() != static_cast<int>(Children().Size()))
     {
-        auto item = m_itemsDataSource.GetAt(i);
-        auto node = winrt::make_self<TreeViewNode>();
-        node->Content(item);
-        winrt::get_self<TreeViewNodeVector>(Children())->InsertAtCore(index, *node);
+        for (int i = index + count - 1; i >= index; i--)
+        {
+            auto item = m_itemsDataSource.GetAt(i);
+            auto node = winrt::make_self<TreeViewNode>();
+            node->Content(item);
+            winrt::get_self<TreeViewNodeVector>(Children())->InsertAt(index, *node, false /* updateItemsSource */);
+        }
     }
 }
 
 void TreeViewNode::OnItemsRemoved(int index, int count)
 {
-    for (int i = 0; i < count; i++)
+    // TreeViewNode and ItemsSource will update each other when data changes.
+    // For ItemsSource -> TreeViewNode changes, m_itemsDataSource.Count() < Children().Size()
+    // We'll remove the node from children collection.
+    // For TreeViewNode -> ItemsSource changes, m_itemsDataSource.Count() == Children().Size()
+    // the node is already removed, we don't want to update TreeViewNode again here.
+    if (m_itemsDataSource.Count() != static_cast<int>(Children().Size()))
```

Comment:
Have you tried readonly itemsource like IEnumerable datasource? I guess you may have weird behavior after drag and drop since treeviewnode didn't get update but UI does.

Reply:
> Have you tried readonly itemsource like IEnumerable datasource? I guess you may have weird behavior after drag and drop since treeviewnode didn't get update but UI does.

This is part of of the ItemsSourceChanged callback, if ItemsSource is read only, we should never reach this code path.

We do need to handle the question you mentioned in the other comment: "what do we do for read only ItemsSource when TreeViewNode changes"

Reply:
Jevan talked to me offline and I think we are ok with the current behavior: we try our best to update ItemsSource when TreeViewNode changes, but if we are unable to update it, we'll keep ItemsSource untouched and only update TreeViewNode.